framework/src/nodalkernels/ADNodalKernel.C

```diff
@@ -0,0 +1,71 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADNodalKernel.h"
+#include "Problem.h"
+#include "SubProblem.h"
+#include "SystemBase.h"
+#include "MooseVariableFE.h"
+#include "Assembly.h"
+
+#include "metaphysicl/raw_type.h"
+
+InputParameters
+ADNodalKernel::validParams()
+{
+  return NodalKernelBase::validParams();
+}
+
+ADNodalKernel::ADNodalKernel(const InputParameters & parameters)
+  : NodalKernelBase(parameters), _u(_var.adDofValues())
+{
+#ifndef MOOSE_GLOBAL_AD_INDEXING
+  mooseError("ADNodalKernels are only supported with global AD indexing");
+#endif
+
+  if (isParamValid("save_in"))
+    paramError("save_in",
+               "ADNodalKernels do not support save_in. Please use the tagging system instead.");
+  if (isParamValid("diag_save_in"))
+    paramError(
+        "diag_save_in",
+        "ADNodalKernels do not support diag_save_in. Please use the tagging system instead.");
+}
+
+void
+ADNodalKernel::computeResidual()
+{
+  if (_var.isNodalDefined())
+  {
+    const auto dof_idx = _var.nodalDofIndex();
+    _qp = 0;
+    auto res = MetaPhysicL::raw_value(computeQpResidual());
+    res *= _var.scalingFactor();
+    _assembly.cacheResidual(dof_idx, res, _vector_tags);
+  }
+}
+
+void
+ADNodalKernel::computeJacobian()
+{
+  if (_var.isNodalDefined())
+  {
+    const auto dof_idx = _var.nodalDofIndex();
+    _qp = 0;
+    const auto res = computeQpResidual();
+    _assembly.processDerivatives(res, dof_idx, _matrix_tags);
+  }
+}
+
+void
+ADNodalKernel::computeOffDiagJacobian(const unsigned int jvar)
+{
+  if (jvar == _var.number())
+    computeJacobian();
+}
```

Comment:
No. We actually only call `computeOffDiagJacobian` for `NodalKernels`. In order to treat `ADNodalKernels` uniquely we would have to store them in their own warehouse. This *is* a no-op unless `jvar == _var.number()`

Reply:
Ah - I see - the compute loop only ever calls the off-diag jac func.  Still feels funny.  If anyone ever called computeOffDiag _only once_ with `i != j` it would silently be a no-op - 😢 . It is unlikely to happen and we don't expect any code to be written that way.  But some day 5 years from now, someone will spend a week trying to figure out why the jacobian isn't being updated like they expect.  So the alternative is to filter the ad kernels and keep them out of the coupling loop and just call computeJacobian once on each of them - and make offdiag either compute jacobian every time or be an error.  That would be my preference.  But I'm not going to be obstinate about it.  You can decide if you think it's worth the trouble.

Reply:
Funny enough - I pinged @lindsayad about something similar. I'm looking into cleaning up the off-diag calls to just off-diag, computeJacobian() for diag, and then an alternate for AD. Will see how it turns out...

Reply:
This is an identical treatment to `ADIntegratedBC` and similar to what is done for `ADKernel`. I think your concerns are valid @rwcarlsen but instead of changing it in this PR I would prefer to open an issue for considering going through the AD objects and splitting them into separate warehouses and changing how they are called... and it sounds like @loganharbour is already working on that issue haha. @loganharbour do you have a ticket for that?